File: examples/canonical-gcp-functions/gcs-trigger-ocr-to-firestore/python/tests/integration/lane-a-minisky/test_handler.py
from pathlib import Path
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "src"))

from handler import Clients, handle_gcs_object


class Firestore:
    def __init__(self) -> None:
        self.claimed: set[str] = set()
        self.metadata: dict[str, dict] = {}

    def claim_generation(self, dedupe_key: str, payload: dict) -> bool:
        if dedupe_key in self.claimed:
            return False
        self.claimed.add(dedupe_key)
        return True

    def write_metadata(self, dedupe_key: str, payload: dict) -> None:
        self.metadata[dedupe_key] = payload


class Vision:
    def ocr(self, bucket: str, name: str, generation: str) -> str:
        return f"ocr:{bucket}/{name}@{generation}"


class PubSub:
    def publish(self, topic: str, payload: dict) -> None:
        return None


def main() -> None:
    clients = Clients(Firestore(), Vision(), PubSub())
    result = handle_gcs_object({"id": "evt-local", "data": {"bucket": "docs", "name": "scan.png", "generation": "1"}}, clients)
    assert result["ok"]


if __name__ == "__main__":
    main()

File: examples/canonical-gcp-functions/gcs-trigger-ocr-to-firestore/go/tests/integration/lane-a-minisky/test_handler.go
package main

import (
	"context"
	"fmt"

	handler "accb.example.gcpfunctions/gcsocrfirestore/src"
)

type fakeClients struct {
	claimed   map[string]bool
	metadata  map[string]map[string]string
	published []map[string]string
}

func (f *fakeClients) ClaimGeneration(ctx context.Context, key string) (bool, error) {
	if f.claimed[key] {
		return false, nil
	}
	f.claimed[key] = true
	return true, nil
}

func (f *fakeClients) RunOCR(ctx context.Context, bucket, name, generation string) (string, error) {
	return fmt.Sprintf("ocr:%s/%s@%s", bucket, name, generation), nil
}

func (f *fakeClients) WriteMetadata(ctx context.Context, key string, payload map[string]string) error {
	f.metadata[key] = payload
	return nil
}

func (f *fakeClients) PublishDownstream(ctx context.Context, payload map[string]string) error {
	f.published = append(f.published, payload)
	return nil
}

func main() {
	clients := &fakeClients{claimed: map[string]bool{}, metadata: map[string]map[string]string{}}
	result, err := handler.Handle(context.Background(), handler.GcsObjectEvent{
		ID:         "evt-lane-a",
		Bucket:     "docs",
		Name:       "scan.png",
		Generation: "8",
	}, clients)
	if err != nil {
		panic(err)
	}
	if result.Duplicate || len(clients.published) != 1 || clients.metadata["docs:scan.png:8"]["status"] != "OCR_COMPLETE" {
		panic("unexpected lane-a GCS OCR effect")
	}
	fmt.Println("lane-a GCS OCR handler invocation passed")
}
